cmd/opampsupervisor/supervisor/supervisor.go

```diff
@@ -0,0 +1,649 @@
+	cfg := s.effectiveConfig.Load().(string)
+	err := s.commander.Stop(context.Background())
+
+	if err != nil {
+		s.logger.Errorf("Could not stop agent process: %v", err)
+	}
+
+	s.writeEffectiveConfigToFile(cfg, s.effectiveConfigFilePath)
+}
+
+func (s *Supervisor) writeEffectiveConfigToFile(cfg string, filePath string) {
+	f, err := os.Create(filePath)
+	if err != nil {
+		s.logger.Errorf("Cannot open effective config file: %v", err)
+	}
+	defer f.Close()
+
+	_, err = f.WriteString(cfg)
+
+	if err != nil {
+		s.logger.Errorf("Cannot write effective config file: %v", err)
+	}
+}
+
+func (s *Supervisor) Shutdown() {
+	s.logger.Debugf("Supervisor shutting down...")
+	if s.commander != nil {
+		err := s.commander.Stop(context.Background())
+
+		if err != nil {
+			s.logger.Errorf("Could not stop agent process: %v", err)
+		}
+	}
+	if s.opampClient != nil {
+		err := s.opampClient.SetHealth(
+			&protobufs.AgentHealth{
+				Healthy: false, LastError: "Supervisor is shutdown",
+			},
+		)
+
+		if err != nil {
+			s.logger.Errorf("Could not report OpAMP client health: %v", err)
+		}
+
+		err = s.opampClient.Stop(context.Background())
+
+		if err != nil {
+			s.logger.Errorf("Could not stop the OpAMP client: %v", err)
+		}
+	}
+}
+
+func (s *Supervisor) onMessage(ctx context.Context, msg *types.MessageData) {
+	configChanged := false
+	if msg.RemoteConfig != nil {
+		s.remoteConfig = msg.RemoteConfig
+		s.logger.Debugf("Received remote config from server, hash=%x.", s.remoteConfig.ConfigHash)
+
+		var err error
+		configChanged, err = s.recalcEffectiveConfig()
+		if err != nil {
+			err = s.opampClient.SetRemoteConfigStatus(&protobufs.RemoteConfigStatus{
+				LastRemoteConfigHash: msg.RemoteConfig.ConfigHash,
+				Status:               protobufs.RemoteConfigStatuses_RemoteConfigStatuses_FAILED,
+				ErrorMessage:         err.Error(),
+			})
+			if err != nil {
+				s.logger.Errorf("Could not report failed OpAMP remote config status: %v", err)
+			}
+		} else {
+			err = s.opampClient.SetRemoteConfigStatus(&protobufs.RemoteConfigStatus{
+				LastRemoteConfigHash: msg.RemoteConfig.ConfigHash,
+				Status:               protobufs.RemoteConfigStatuses_RemoteConfigStatuses_APPLIED,
+			})
+			if err != nil {
+				s.logger.Errorf("Could not report applied OpAMP remote config status: %v", err)
+			}
+		}
+	}
+
+	if msg.OwnMetricsConnSettings != nil {
+		configChanged = s.setupOwnMetrics(ctx, msg.OwnMetricsConnSettings) || configChanged
+	}
+
+	if msg.AgentIdentification != nil {
+		newInstanceID, err := ulid.Parse(msg.AgentIdentification.NewInstanceUid)
+		if err != nil {
+			s.logger.Errorf(err.Error())
+		}
+
+		s.logger.Debugf("Agent identify is being changed from id=%v to id=%v",
+			s.instanceID.String(),
+			newInstanceID.String())
+		s.instanceID = newInstanceID
+
+		// TODO: update metrics pipeline by altering configuration and setting
+		// the instance id when Collector implements https://github.com/open-telemetry/opentelemetry-collector/pull/5402.
```

Comment:
https://github.com/open-telemetry/opentelemetry-collector/pull/5402 is already implemented and I think we just need to set `configChanged` to true to trigger the instance id update via config update.